File: Programmers/Algorithm/Hash/failed_runner.py
# https://programmers.co.kr/learn/courses/30/lessons/42576?language=python3 [완주하지 못한 선수]
# HASH

def solution(participant, completion):
    answer = ''
    participant.sort()
    completion.sort()
    completion.append("1")
    
    for p in range(len(participant)):
        if participant[p] != completion[p]:
            return participant[p]

# collections과 Counter를 이용한 풀이
import collections


def solution(participant, completion):
    answer = collections.Counter(participant) - collections.Counter(completion)
    return list(answer.keys())[0] #list로 형변환 중요

File: Programmers/Algorithm/Hash/failed_runner.js
function solution(participant, completion) {
  let answer = '';
  let runners = new Map();

  participant.forEach((runner) => {
    runners.set(runner, (runners.get(runner) || 0) + 1);
  });

  completion.forEach((completed) => {
    runners.set(completed, (runners.get(completed) || 0) - 1);
    if (runners.get(completed) === 0) runners.delete(completed);
  });

  for (let [name, val] of runners) {
    return name;
  }
}

console.log(solution(['mislav', 'stanko', 'mislav', 'ana'], ['stanko', 'ana', 'mislav']));
